README

This is a modified version of RTG poll that is available on sourceforge at http://rtg.sourceforge.net/

This version has stripped out any reference to MySQL and replaced the logging of collected SNMP data to a database with integration to Ganglia for data retention via RRD's and graphing.




NEWS
rtg-0.7.4 (2003-10-02)
---------
* Added fast target hashing in rtgpoll.
* Added plotting of interface aggregates into rtgplot.
* Added multiple simultaneous interface plotting into rtgplot.
* Added y-scaling by interface speed support into rtgplot.  
* Added arbitrary image and border sizing into rtgplot.
* Added using autoconf defined --prefix as search path for rtg.conf.
* Added graph fill of area under plot lines into rtgplot.
* Added rtgpoll option (-m) to skip checking for multiple instances.
* Added rtgpoll option (-z) to force database insertion for zero deltas.
* Added debug file pointer; allows writing debug info to file while 
   running rtgplot via CGI.
* Added contrib directory to top-level distribution tree.
* Added libz probing in configure; cleaned up probing logic.
* Changed autoconf SNMP library probing order to detect Net-SNMP first.
* Changed autoconf and code to use strtoll() if possible.
* Audit code for potential buffer overflows; modified all code to use 
   strncpy(), snprintf, and sscan with width specifier.
* Modified all php to use full tags instead of short tags.
* Thread code cleaned and logic improved.
* Fixed bug in rtgplot that incorrectly counted skewed measurements
   in rate max and average.
* From: Chris Brotsos
   Support for per-target maximum polled values.
* From: Chris Brotsos
   Patch to allow ASN_INTEGER/ASN_OPAQUE type responses. 
* From: Shayne Smyth
   Patch to allow ASN_TIMETICKS type responses.
* From: Chris Brotsos
   Patch to implement a PID file and additional signal catching.
* From: Joost Lek
   Support for Nth percentile lines in rtgplot.
* From: Chris Brotsos
   Patch to allow rtgplot to determine last database value when last
   delta occurs outside of queried range.
* From: Michal Szczotka
   Patch to fix division by zero in 95.pl and report.pl.
* From: Brian T. O'Neill
   Patch for extending YPLOT_AREA with GAUGE data lines to allow visibility. 
* From: Brian T. O'Neill
   Patch for allowing configurable HighSkewSlop and LowSkewSlop (See README).
* From: Brian T. O'Neill
   Patch to implement factor logic with GAUGE data.
* From: Brian T. O'Neill
   Modularized rtgtargmkr included in contrib.
* From: Brian T. O'Neill
   Patch to properly read large interface speeds from database.


rtg-0.7.3 (2003-02-08)
---------
* Fixed bug in poller() where last_value object was improperly updated
   when polled device did not respond.
* Fixed rtgplot.cgi to support apache2.
* Fixed read_rtg_config() to handle comments and arbitrary ordering.
* Fixed NULL pointer crash in rtgsnmp.c snmp_free_pdu().
* Fixed rtgplot to detect, compensate and properly plot missed polls.
* Fixed rtgplot to time-align samples between multiple data sources
   when scalex is defined and the data (in database) for one begins
   before the other.
* Fixed FreeBSD autoconf support.
* Added gauge (absolute SNMP counter) support.
* Added config directive to specify SNMP version (1 or 2).
* Added config directive to specify number of threads.
* Added timestamp to poll begin/complete messages.
* Added poll round counter to poll complete messages.
* Added shading to rtgplot.c
* Added red arrow showing direction of growth to rtgplot graphs.
* Added autoconf detection for libz if required by MySQL.
* Added 95.php, 95th percentile report PHP web script.
* Added view.php, an MRTG-style day/week/month display for RTG in PHP.
* Modified all .pl and .php scripts to find and read from the 
   master rtg.conf first, then fall back to defaults.
* Modified rtgtargmkr.pl to read counter bit size from routers
   configuration file to allow size to be specified per router.
* Modified sleepy() routine to use usleep calls making the sleep
   interval exact and the period between polls accurate.
* Modified rtgtargmkr.pl to write out IPs instead of hostnames so
   that DNS failures do not impact collection.
* Aligned rate text in rtgplot.c properly.
* Updated Simon Leinen's included SNMP_Session to version 0.94 (used 
   by rtgtargmkr.pl).
* From: Jon Dustin
   Patch to rtgtargmkr.pl to perform case insensitive checking of
   reserved interfaces.
* From: Stephen Fromm
   Patch to configure to deal with multiple MySQL installs.


rtg-0.7.2 (2002-10-04)
---------
* Fixed report.pl to handle interfaces with zero speed gracefully.
* Rewrote autoconf to detect either UCD-SNMP or Net-SNMP.
* Fixed incompatibility with ucd/net-snmp versions >= 5.x.  Poller now 
   fixed to use only 5.x SNMP functions.
* Fixed bad SQL update statement in rtgtargmkr.pl
* Made rtgtargmkr.pl Cisco Catalyst aware.
* Added time range argument to 95th percentile report
* From: David A. Snodgrass
   Patch to rtgtargmkr.pl to handle interface descriptions with quotes.


rtg-0.7.1 (2002-08-28)
---------
* Fixed bad overflow with 64-bit SNMP counters.
* Fixed database schema to accommodate long interface names.
* Fixed rtgtargmkr.pl router file parsing.
* Fixed rtgtargmkr.pl bug handling descriptions with trailing whitespace.
* From: Adam Rothschild
   Patch to report.pl to specify time periods.
* From: David A. Snodgrass
   Changed createdb MySQL permissions to be more restrictive.
* From: Omachonu Ogali
   Patch to rtgtargmkr.pl to support per-device communities.


rtg-0.7 (2002-08-07)
-------
* Major rewrite of poller to be entirely multi-threaded.  Assumes
   full pthreads support and MySQL thread-safe client support.  
* Autoconf fixes to support FreeBSD -pthreads, Solaris thread
   compiler flags, checks for long longs.
* Changed insert date to NOW() function to reflect SQL server's current
   time.  This removes any synchronization issues in the RTG pollers.
* Fixed SIGHUP initiated memory leak.
* Cleaned up installation and installation scripts.
* Added Perl DBI reports.
* Added Perl target file maintainer script.
* From: David A. Snodgrass
   Patch to createdb to include autoconf MySQL path


rtg-0.6 (2002-07-23)
-------
* Added randomization of target list to "even out" SNMP query load on 
   network devices.

etc/rtg.php
<?php
  include('./common.php');

  print "<HTML>\n<!-- RTG Version $VERSION -->\n<HEAD>\n";

  /* Connect to RTG MySQL Database */
  $dbc=@mysql_connect ($host, $user, $pass) or
  $dbc=@mysql_connect ("$host:/var/lib/mysql/mysql.sock", $user, $pass) or 
     die ("MySQL Connection Failed, Check Configuration.");
  mysql_select_db($db,$dbc);

 if ($PHP_SELF == "") {
   $PHP_SELF = "rtg.php";
   $errors = $_GET['errors'];
   $scalex = $_GET['scalex'];
   $scaley = $_GET['scaley'];
   $aggr = $_GET['aggr'];
   $percentile = $_GET['percentile'];
   $nth = $_GET['nth'];
   $xplot = $_GET['xplot'];
   $yplot = $_GET['yplot'];
   $borderb = $_GET['borderb'];
   $iid = $_GET['iid'];
   $rid = $_GET['rid'];
   $syear = $_GET['syear'];
   $eyear = $_GET['eyear'];
   $smonth = $_GET['smonth'];
   $emonth = $_GET['emonth'];
   $sday = $_GET['sday'];
   $eday = $_GET['eday'];
   $shour = $_GET['shour'];
   $ehour = $_GET['ehour'];
   $smin = $_GET['smin'];
   $emin = $_GET['emin'];
 }

  # Determine router, interface names as necessary
  if ($rid && $iid) {
    $selectQuery="SELECT a.name, a.description, a.speed, b.name AS router FROM interface a, router b WHERE a.rid=b.rid AND a.rid=$rid AND a.id=$iid[0]";
    $selectResult=mysql_query($selectQuery, $dbc);
    $selectRow=mysql_fetch_object($selectResult);
    $interfaces = mysql_num_rows($selectResult);
    $name = $selectRow->name;
    $description = $selectRow->description;
    $speed = ($selectRow->speed)/1000000;
    $router = $selectRow->router;
  } else if ($rid && !$iid) {
    $selectQuery="SELECT name AS router from router where rid=$rid";
    $selectResult=mysql_query($selectQuery, $dbc);
    $selectRow=mysql_fetch_object($selectResult);
    $router = $selectRow->router;
  }

  # Generate Title 
  echo "<TITLE>RTG: ";
  if ($rid && $iid) echo "$router: $name";
  else if ($rid && !$iid) echo "$router";
  echo "</TITLE>\n";
?>

</HEAD>
<BODY BACKGROUND="rtgback.png" BGCOLOR="ffffff">
<A HREF="http://rtg.sourceforge.net" ALT="[RTG Home Page]">
<IMG SRC="rtg.png" BORDER="0">
</A>
<P>

<?php
echo "<FORM ACTION=\"$PHP_SELF\" METHOD=\"get\">\n";

if (!$rid && !$iid) {
 echo "<SELECT NAME=\"rid\" SIZE=10>\n";
 $selectQuery="SELECT DISTINCT name, rid FROM router ORDER BY name";
 $selectResult=mysql_query($selectQuery, $dbc);
 while ($selectRow=mysql_fetch_object($selectResult)){
    echo "<OPTION VALUE=\"$selectRow->rid\">$selectRow->name\n";
 }
 echo "</SELECT>\n";
}

if ($smonth != "" && $iid == "") {
 echo "<BLOCKQUOTE><FONT SIZE=+1><STRONG>\n";
 echo "Please select an interface!\n";
 echo "</FONT></STRONG></BLOCKQUOTE>\n";
}

if ($rid && !$iid) {
  echo "<SELECT NAME=\"rid\" SIZE=10>\n"; 
  echo "<OPTION SELECTED VALUE=\"$rid\">$router\n";
  echo "</SELECT>\n";

  echo "<SELECT MULTIPLE NAME=\"iid[]\" SIZE=10>\n"; 
  $selectQuery="SELECT id, name, description FROM interface WHERE rid=$rid ORDER BY name";
  $selectResult=mysql_query($selectQuery, $dbc);
  while ($selectRow=mysql_fetch_object($selectResult)){
     echo "<OPTION VALUE=\"$selectRow->id\">$selectRow->name ($selectRow->description)\n";
  }
  echo "</SELECT>\n";

  echo "<P><TABLE>\n";
  echo "<TD>From: \n";
  echo "<TD><INPUT TYPE=TEXT NAME=\"smonth\" SIZE=3 MAXLENGTH=2 VALUE=\"";
  print (date("m"));
  echo "\"> / \n";
  echo "<INPUT TYPE=TEXT NAME=\"sday\" SIZE=3 MAXLENGTH=2 VALUE=\"";
  print (date("d"));
  echo "\"> / \n";
  echo "<INPUT TYPE=TEXT NAME=\"syear\" SIZE=5 MAXLENGTH=4 VALUE=\"";
  print (date("Y"));
  echo "\">\n";
  echo "<TD><INPUT TYPE=TEXT NAME=\"shour\" SIZE=3 MAXLENGTH=2 VALUE=\"00\">:\n";
  echo "<INPUT TYPE=TEXT NAME=\"smin\" SIZE=3 MAXLENGTH=2 VALUE=\"00\">\n";

  echo "<TR><TD>To: \n";
  echo "<TD><INPUT TYPE=TEXT NAME=\"emonth\" SIZE=3 MAXLENGTH=2 VALUE=\"";
  print (date("m"));
  echo "\"> / \n";
  echo "<INPUT TYPE=TEXT NAME=\"eday\" SIZE=3 MAXLENGTH=2 VALUE=\"";
  print (date("d"));
  echo "\"> / \n";
  echo "<INPUT TYPE=TEXT NAME=\"eyear\" SIZE=5 MAXLENGTH=4 VALUE=\"";
  print (date("Y"));
  echo "\">\n";
  echo "<TD><INPUT TYPE=TEXT NAME=\"ehour\" SIZE=3 MAXLENGTH=2 VALUE=\"23\">:\n";
  echo "<INPUT TYPE=TEXT NAME=\"emin\" SIZE=3 MAXLENGTH=2 VALUE=\"59\">\n";
  echo "<TR><TD>Aggregate Interfaces: \n";
  echo "<TD><INPUT TYPE=CHECKBOX NAME=\"aggr\">\n";
  echo "<TR><TD><INPUT TYPE=TEXT NAME=\"nth\" SIZE=2 MAXLENGTH=2 VALUE=\"95\">th Percentile: \n";
  echo "<TD><INPUT TYPE=CHECKBOX NAME=\"percentile\">\n";
  echo "<TR><TD>Fit to Data: \n";
  echo "<TD><INPUT TYPE=CHECKBOX NAME=\"scalex\">\n";
  echo "<TR><TD>Fit to Speed: \n";
  echo "<TD><INPUT TYPE=CHECKBOX NAME=\"scaley\">\n";
  echo "<TR><TD>X Size: \n";
  echo "<TD><INPUT TYPE=TEXT NAME=\"xplot\" SIZE=3 VALUE=\"500\">\n";
  echo "<TR><TD>Y Size: \n";
  echo "<TD><INPUT TYPE=TEXT NAME=\"yplot\" SIZE=3 VALUE=\"150\">\n";
  echo "<TR><TD>BorderB: \n";
  echo "<TD><INPUT TYPE=TEXT NAME=\"borderb\" SIZE=3 VALUE=\"70\">\n";
  echo "</TABLE>\n";
}

if (($bt || $smonth) && $iid) { 

  /* Format into GNU date syntax */
  if ($bt == "") {
    $bt = strtotime("$syear-$smonth-$sday $shour:$smin:00");
    $et = strtotime("$eyear-$emonth-$eday $ehour:$emin:59");
  }

  $range="dtime>FROM_UNIXTIME($bt) AND dtime<=FROM_UNIXTIME($et)";
  $range="$range AND id=$iid[0]";

  $selectQuery="SELECT description, name, speed FROM interface WHERE rid=$rid AND id=$iid[0]";
  $selectResult=mysql_query($selectQuery, $dbc);
  $selectRow=mysql_fetch_object($selectResult);
  echo "<TABLE BORDER=0>\n";
  echo "<TD><I>Device</I>:</TD><TD>$router ($rid)</TD><TR>\n";
  echo "<TD><I>Interface</I>:</TD><TD>$selectRow->name ($iid[0])</TD><TR>\n";
  printf ("<TD><I>Speed</I>:</TD><TD>%2.3f Mbps</TD><TR>\n", $selectRow->speed / 1000000);
  echo "<TD><I>Description</I>:</TD><TD>$selectRow->description</TD><TR>\n";
  print strftime("<TD><I>Period</I>:</TD><TD>%m/%d/%Y %H:%M - ", $bt);
  print strftime("%m/%d/%Y %H:%M</TD>\n", $et);
  echo "</TABLE>\n";
  echo "<P>\n";
 
  #$selectQuery="SELECT DISTINCT id FROM ifInOctets_$rid WHERE $range";
  #$selectResult=mysql_query($selectQuery, $dbc);
  #if (mysql_num_rows($selectResult) <= 0) {
  #   print "<BR>No Data Found on Interface for Given Range.<BR>\n";
  #}
  #else {
    $args = "t1=ifInOctets_$rid&t2=ifOutOctets_$rid&begin=$bt&end=$et&units=bits/s&factor=8";
    foreach ($iid as $value) {
      $args="$args&iid=$value";
    }
    if ($scalex) $args = "$args&scalex=yes";
    if ($scaley) $args = "$args&scaley=yes";
    if ($xplot) $args = "$args&xplot=$xplot";
    if ($yplot) $args = "$args&yplot=$yplot";
    if ($borderb) $args = "$args&borderb=$borderb";
    if ($aggr) $args = "$args&aggr=yes";
    if ($percentile) $args = "$args&percentile=$nth";
    print "<IMG SRC=rtgplot.cgi?$args><BR>\n";
    $args = "t1=ifInUcastPkts_$rid&t2=ifOutUcastPkts_$rid&begin=$bt&end=$et&units=pkts/s";
    foreach ($iid as $value) {
      $args="$args&iid=$value";
    }
    if ($scalex) $args = "$args&scalex=yes";
    if ($xplot) $args = "$args&xplot=$xplot";
    if ($yplot) $args = "$args&yplot=$yplot";
    if ($borderb) $args = "$args&borderb=$borderb";
    if ($aggr) $args = "$args&aggr=yes";
    if ($percentile) $args = "$args&percentile=$nth";
    print "<IMG SRC=rtgplot.cgi?$args><BR>\n";
    if ($errors) 
       print "<IMG SRC=rtgplot.cgi?t1=ifInErrors_$rid&begin=$bt&end=$et&units=errors&impulses=yes>\n";
#  }
} 

mysql_close($dbc);
?>

<P>
<INPUT TYPE="SUBMIT" VALUE="Ok">
</FORM>
<BR>
<HR>
<FONT FACE="Arial" SIZE="2">
<?php
 print "<A HREF=\"http://rtg.sourceforge.net\">RTG</A> Version $VERSION</FONT>";
?>
</BODY>
</HTML>
